percona/controller/pgcluster/backup.go

```diff
@@ -11,10 +11,83 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
+	"github.com/percona/percona-postgresql-operator/internal/logging"
 	"github.com/percona/percona-postgresql-operator/internal/naming"
+	"github.com/percona/percona-postgresql-operator/percona/controller"
+	"github.com/percona/percona-postgresql-operator/percona/pgbackrest"
 	v2 "github.com/percona/percona-postgresql-operator/pkg/apis/pgv2.percona.com/v2"
 )
 
+func (r *PGClusterReconciler) reconcileBackups(ctx context.Context, cr *v2.PerconaPGCluster) error {
+	if err := r.reconcileBackupJobs(ctx, cr); err != nil {
+		return errors.Wrap(err, "reconcile backup jobs")
+	}
+
+	if err := r.cleanupOutdatedBackups(ctx, cr); err != nil {
+		return errors.Wrap(err, "cleanup outdated backups")
+	}
+
+	return nil
+}
+
+func (r *PGClusterReconciler) cleanupOutdatedBackups(ctx context.Context, cr *v2.PerconaPGCluster) error {
+	log := logging.FromContext(ctx)
+
+	if cr.Status.State != v2.AppStateReady {
+		return nil
+	}
+
+	for _, repo := range cr.Spec.Backups.PGBackRest.Repos {
+		var info pgbackrest.InfoOutput
+
+		pbList, err := listPGBackups(ctx, r.Client, cr, repo.Name)
+		if err != nil {
+			return errors.Wrap(err, "list pg-backups")
+		}
+		if len(pbList) == 0 {
+			continue
+		}
+
+		readyPod, err := controller.GetReadyInstancePod(ctx, r.Client, cr.Name, cr.Namespace)
+		if err != nil {
+			return errors.Wrap(err, "get ready instance pod")
+		}
+		info, err = pgbackrest.GetInfo(ctx, readyPod, repo.Name)
+		if err != nil {
+			if errors.Is(err, pgbackrest.ErrNoValidBackups) {
+				log.Error(err, "Failed to get pgbackrest info", "repo", repo.Name)
+				continue
+			}
+			return errors.Wrap(err, "get pgBackRest info")
+		}
+
+		for _, pgBackup := range pbList {
+			if pgBackup.Status.State != v2.BackupSucceeded || pgBackup.CompareVersion("2.4.0") < 0 {
+				continue
+			}
+
+			backupPresent := false
+			for _, info := range info {
+				for _, backupInfo := range info.Backup {
+					if backupInfo.Annotation[v2.PGBackrestAnnotationJobName] == pgBackup.Status.JobName {
+						backupPresent = true
+						break
+					}
+				}
+			}
+			if backupPresent {
+				continue
+			}
```

Comment:
i don't understand why do we have this check

Reply:
@pooknull correct me if I'm wrong, but this means that pgbackrest didn't delete the backup (meaning it didn't expire) so we shouldn't delete backup object associated with it. If the backup is expired and removed by pgbackrest only then we need to remove associated backup object.

Reply:
@inelpandzic you're totally correct